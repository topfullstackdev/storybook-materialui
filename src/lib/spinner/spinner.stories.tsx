import { Meta, Story } from '@storybook/react';

import { Spinner, SpinnerProps } from './spinner';

export default {
  component: Spinner,
  title: 'Spinner',
  args: {}, // Common args
} as Meta<SpinnerProps>;

function Template(props: SpinnerProps) {
  return <Spinner {...props} />;
}

export const Primary: Story<SpinnerProps> = Template.bind({});
