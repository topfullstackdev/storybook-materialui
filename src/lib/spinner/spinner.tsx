import { css, keyframes } from '@emotion/react';
import { useMemo } from 'react';

import { useTheme } from '@bai/utils/theme';

const margins = {
  all: '100px auto',
  right: '0px 75px 0px 0px',
  none: '15px auto',
};

export type Margin = keyof typeof margins;

export interface SpinnerProps {
  margin?: Margin;
}

const stretchDelay = keyframes`
  from, 0%, 40%, 100%, to {
    transform: scaleY(0.4);
  }
  20% {
    transform: scaleY(1.0);
  }
`;

export function Spinner({ margin = 'all' }: SpinnerProps) {
  const theme = useTheme();
  /* eslint-disable react/no-array-index-key */
  const bars = useMemo(
    () =>
      new Array(5).fill(0).map((el, i) => (
        <div
          key={i}
          css={css`
            background-color: ${theme?.branding?.primaryColor};
            height: 100%;
            width: 6px;
            margin: 0 2px;
            display: inline-block;
            animation: ${stretchDelay} 1.2s infinite ease-in-out;
            animation-delay: ${-1.2 + i / 10}s;
          `}
        />
      )),
    [theme?.branding?.primaryColor],
  );
  /* eslint-enable react/no-array-index-key */
  return (
    <div
      css={css`
        margin: ${margins[margin]};
        width: 50px;
        height: 40px;
        text-align: center;
        font-size: 10px;
        z-index: 20;
      `}
    >
      {bars}
    </div>
  );
}

export default Spinner;
