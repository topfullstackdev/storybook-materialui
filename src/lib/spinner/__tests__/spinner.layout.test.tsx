// Uses snapshot to test layout rendering
import { ThemeProvider } from '@mui/material/styles';
import { render } from '@testing-library/react';

import { createTheme } from '@bai/utils/theme';

import { SpinnerProps, Spinner as SpinnerRaw } from '../spinner';

const theme = createTheme({
  branding: {
    primaryColor: '#B22028',
  },
});

describe('Spinner', () => {
  function Spinner(props: SpinnerProps) {
    return (
      <ThemeProvider theme={theme}>
        <SpinnerRaw {...props} />
      </ThemeProvider>
    );
  }

  // The only prop affects CSS, which doesn't render in the snapshot
  it('should render successfully', () => {
    const { container } = render(<Spinner />);
    expect(container).toMatchSnapshot();
  });
});
