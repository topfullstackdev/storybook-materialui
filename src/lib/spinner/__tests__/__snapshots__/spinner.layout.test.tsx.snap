// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should render successfully 1`] = `
@keyframes animation-0 {
  from, 0%, 40%, 100%, to {
    -webkit-transform: scaleY(0.4);
    -moz-transform: scaleY(0.4);
    -ms-transform: scaleY(0.4);
    transform: scaleY(0.4);
  }

  20% {
    -webkit-transform: scaleY(1.0);
    -moz-transform: scaleY(1.0);
    -ms-transform: scaleY(1.0);
    transform: scaleY(1.0);
  }
}

@keyframes animation-0 {
  from, 0%, 40%, 100%, to {
    -webkit-transform: scaleY(0.4);
    -moz-transform: scaleY(0.4);
    -ms-transform: scaleY(0.4);
    transform: scaleY(0.4);
  }

  20% {
    -webkit-transform: scaleY(1.0);
    -moz-transform: scaleY(1.0);
    -ms-transform: scaleY(1.0);
    transform: scaleY(1.0);
  }
}

@keyframes animation-0 {
  from, 0%, 40%, 100%, to {
    -webkit-transform: scaleY(0.4);
    -moz-transform: scaleY(0.4);
    -ms-transform: scaleY(0.4);
    transform: scaleY(0.4);
  }

  20% {
    -webkit-transform: scaleY(1.0);
    -moz-transform: scaleY(1.0);
    -ms-transform: scaleY(1.0);
    transform: scaleY(1.0);
  }
}

@keyframes animation-0 {
  from, 0%, 40%, 100%, to {
    -webkit-transform: scaleY(0.4);
    -moz-transform: scaleY(0.4);
    -ms-transform: scaleY(0.4);
    transform: scaleY(0.4);
  }

  20% {
    -webkit-transform: scaleY(1.0);
    -moz-transform: scaleY(1.0);
    -ms-transform: scaleY(1.0);
    transform: scaleY(1.0);
  }
}

@keyframes animation-0 {
  from, 0%, 40%, 100%, to {
    -webkit-transform: scaleY(0.4);
    -moz-transform: scaleY(0.4);
    -ms-transform: scaleY(0.4);
    transform: scaleY(0.4);
  }

  20% {
    -webkit-transform: scaleY(1.0);
    -moz-transform: scaleY(1.0);
    -ms-transform: scaleY(1.0);
    transform: scaleY(1.0);
  }
}

.emotion-0 {
  margin: 100px auto;
  width: 50px;
  height: 40px;
  text-align: center;
  font-size: 10px;
  z-index: 20;
}

.emotion-1 {
  background-color: #B22028;
  height: 100%;
  width: 6px;
  margin: 0 2px;
  display: inline-block;
  -webkit-animation: animation-0 1.2s infinite ease-in-out;
  animation: animation-0 1.2s infinite ease-in-out;
  -webkit-animation-delay: -1.2s;
  animation-delay: -1.2s;
}

.emotion-2 {
  background-color: #B22028;
  height: 100%;
  width: 6px;
  margin: 0 2px;
  display: inline-block;
  -webkit-animation: animation-0 1.2s infinite ease-in-out;
  animation: animation-0 1.2s infinite ease-in-out;
  -webkit-animation-delay: -1.0999999999999999s;
  animation-delay: -1.0999999999999999s;
}

.emotion-3 {
  background-color: #B22028;
  height: 100%;
  width: 6px;
  margin: 0 2px;
  display: inline-block;
  -webkit-animation: animation-0 1.2s infinite ease-in-out;
  animation: animation-0 1.2s infinite ease-in-out;
  -webkit-animation-delay: -1s;
  animation-delay: -1s;
}

.emotion-4 {
  background-color: #B22028;
  height: 100%;
  width: 6px;
  margin: 0 2px;
  display: inline-block;
  -webkit-animation: animation-0 1.2s infinite ease-in-out;
  animation: animation-0 1.2s infinite ease-in-out;
  -webkit-animation-delay: -0.8999999999999999s;
  animation-delay: -0.8999999999999999s;
}

.emotion-5 {
  background-color: #B22028;
  height: 100%;
  width: 6px;
  margin: 0 2px;
  display: inline-block;
  -webkit-animation: animation-0 1.2s infinite ease-in-out;
  animation: animation-0 1.2s infinite ease-in-out;
  -webkit-animation-delay: -0.7999999999999999s;
  animation-delay: -0.7999999999999999s;
}

<div>
  <div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    />
    <div
      class="emotion-2"
    />
    <div
      class="emotion-3"
    />
    <div
      class="emotion-4"
    />
    <div
      class="emotion-5"
    />
  </div>
</div>
`;
